[jsk_perception/point_pose_extractor] Use DiagnosticNodelet to improve subscribing

nakane11 opened the pull request:
### Problem
- During the fetch's kitchen demo, this node in trashbin_occupancy_detector continues logging `"wait for subscriberes .."`
- If `SetTemplate` service is requested before `imagefeature_cb` is called, `cv::warpPerspective` raises an error because `PinholeCameraModel` has not been set
- `cv::findHomography()` could returns empty matrix
https://docs.opencv.org/3.3.0/d9/d0c/group__calib3d.html#ga4abc2ece9fab9398f2e560d53c8c9780

I modified point_pose_extractor to inherit DiagnosticNodelet thanks to @iory and added some processing to avoid the errors above.

iory commented on jsk_perception/CMakeLists.txt:
```diff
@@ -234,6 +234,7 @@ jsk_add_nodelet(src/sparse_image_decoder.cpp "jsk_perception/SparseImageDecoder"
 jsk_add_nodelet(src/color_histogram.cpp "jsk_perception/ColorHistogram" "color_histogram")
 jsk_add_nodelet(src/background_substraction_nodelet.cpp "jsk_perception/BackgroundSubstraction" "background_substraction")
 jsk_add_nodelet(src/fisheye_to_panorama.cpp "jsk_perception/FisheyeToPanorama" "fisheye_to_panorama")
+jsk_add_nodelet(src/point_pose_extractor.cpp "jsk_perception/PointPoseExtractor" "point_pose_extractor")
```
Could you add test?
https://github.com/jsk-ros-pkg/jsk_recognition/blob/0de1e9ac4ec9ea27c548d835c9ca5e94a6101808/jsk_perception/CMakeLists.txt#L468-L582

nakane11 replied:
I added the test in 4a7aeda

iory replied:
Nice!

nakane11 commented:
I replaced images for sample because previously object dtection did not succeed.
It enabled point_pose_extractor.test to check if object detection is published.

Should I split these commits?

iory commented:
> Should I split these commits?

You don't have to.

Could you update the document for other users?
https://github.com/jsk-ros-pkg/jsk_recognition/blob/master/doc/jsk_perception/nodes/point_pose_extractor.md
For example, how to use the samples you made, etc.

nakane11 commented:
Sorry for the late update. 
I edited the document.

Affonso-Gui commented on doc/jsk_perception/nodes/point_pose_extractor.md:
```diff
@@ -8,7 +8,8 @@
 - `/ImageFeature0D` (`posedetection_msgs::ImageFeature0D`)
 
-  Image, camera and tempalte feature information. 
+  Image, camera and tempalte feature information.
```
template

nakane11 replied:
Thank you. I fixed it.

nakane11 commented:
Template added from mouse callback is saved as `filename`. 
https://github.com/jsk-ros-pkg/jsk_recognition/blob/1704f9b974cf9901e8feb64639c32267e8bde94c/jsk_perception/include/jsk_perception/point_pose_extractor.h#L659-L667

1. `fname.stem()` and `fname.extension()` put string between " ", so when I input `hoge.png`, `ss` contained `"hoge"_wrap".png"`. I modified it to contain `hoge_wrap.png`.
2. Previously pictures are saved in .ros directry because `fname.stem()` returns filename without path. I attatched `fname.parent_path()` to stem. Now we can save a template as `/tmp/hoge.png`.

k-okada commented on jsk_perception/plugins/nodelet/libjsk_perception.xml:
```diff
@@ -290,6 +290,9 @@
   </class>
+  <class name="jsk_perception/PointPoseExtractor"
+         type="jsk_perception::PointPoseExtractor"
+         base_class_type="nodelet::Nodelet">
```
@nakane11 are we missing `</class>` tag?

nakane11 replied:
I'm sorry, I overlooked that. 
19d2e95 added the end tag.
